### ISSUE TITLE
Spread of optional properties can result in unexpected runtime errors

### ISSUE THREAD

**Issue by issacgerges** [microsoft/TypeScript#44432]
# Bug Report
I believe this may just be a gap in how optional interface properties work under the hood, but it's possible to (using spread or assign) produce values that don't match the type systems expectations. You can see that in this [example](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&ts=3.3.3&ssl=7&ssc=3&pln=1&pc=1&strict=true&noPropertyAccessFromIndexSignature=true&noImplicitOverride=true&noFallthroughCasesInSwitch=true#code/JYOwLgpgTgZghgYwgAgPIAczAPYjgGwDFttkBvAKGWuRhIH4AuZAJQgWygBMAeAZzBRQAcwA0yEAFcAtgCNoAPgoBfChQ4gBySSC4QYoCF2LZmGLLgInkAXnJUadU9t37DXFQG41GrX3RQEHBctvY0tCTMZMqiDtQAdIk6egYgRiZePrhacHx8wMJpIXaosgBW7GDxufmFABTR4mROUcoxLinuJgCU3hT+gcHxTgDaAORwYwC6nsgA9HPIAMJwICDYYMiDIQHY6NBgAJ7IE2PI2DAdbkUUNQVFwyTjkzPziytrG1tBO1B7B8dTudLslrkY1AtkAB1AAWcE2wD4yDh6H2IBEVyg+EOIjUoEgsEQKDYAEdJMBAsYSGFHJFWOxOLwBEIQGIJDJ5FAlKoKEd9sgAJLgaB8So4EChcziqzUgBk9LJFPSJCymk2cGYQoJooQFgldko4RaV1S4OU3iAA), but in short:  Spread/assign are internally check `[[GetOwnProperty]]` to overwrite a value, so `undefined` properties can overwrite ones with concrete values. This results in values that don't represent the calculated Intersection.

I think this is caused by how optional properties are represent internally as (at times) `T | undefined`. Although this holds true for object lookup `foo[prop]`, it does mean that we can create some incorrect values such as this.

```
interface OptionalFoo {
    foo?: Record<string, number>
}

const undefinedFoo: OptionalFoo = {
    foo: undefined
};
```

Resulting in errors like this

```
const spread = {
    foo: {},
    ...undefinedFoo
};

const assigned = Object.assign({}, {foo: {}}, undefinedFoo);

// TS has typed these as `number` but at runtime will produce "Cannot read property 'a' of undefined"
spread.foo['a'];
assigned.foo['a'];

```

### 🔎 Search Terms

interface types, optional properties, intersection, spread, assign

### 🕗 Version & Regression Information

This is the behavior in every version I tried, and I reviewed the FAQ for entries about optional properties and spread/assign

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true&ts=3.3.3&ssl=7&ssc=3&pln=1&pc=1&strict=true&noPropertyAccessFromIndexSignature=true&noImplicitOverride=true&noFallthroughCasesInSwitch=true#code/JYOwLgpgTgZghgYwgAgPIAczAPYjgGwDFttkBvAKGWuRhIH4AuZAJQgWygBMAeAZzBRQAcwA0yEAFcAtgCNoAPgoBfChQ4gBySSC4QYoCF2LZmGLLgInkAXnJUadU9t37DXFQG41GrX3RQEHBctvY0tCTMZMqiDtQAdIk6egYgRiZePrhacHx8wMJpIXaosgBW7GDxufmFABTR4mROUcoxLinuJgCU3hT+gcHxTgDaAORwYwC6nsgA9HPIAMJwICDYYMiDIQHY6NBgAJ7IE2PI2DAdbkUUNQVFwyTjkzPziytrG1tBO1B7B8dTudLslrkY1AtkAB1AAWcE2wD4yDh6H2IBEVyg+EOIjUoEgsEQKDYAEdJMBAsYSGFHJFWOxOLwBEIQGIJDJ5FAlKoKEd9sgAJLgaB8So4EChcziqzUgBk9LJFPSJCymk2cGYQoJooQFgldko4RaV1S4OU3iAA)

### 💻 Code

(see in description)

### 🙁 Actual behavior

The type system does not report any errors although the code is guaranteed to produce an error

### 🙂 Expected behavior

Type system should prevent an optional property from being assigned to `undefined` unless it was defined as `| undefined` or the Intersection should contain `| undefined`

**Comment by fatcerberus** [comment]
This is the type of thing `--strictOptionalProperties` was recently introduced to deal with.

**Comment by RyanCavanaugh** [comment]
With `strictOptionalProperties`, there's an error on line 6:
```ts
interface OptionalFoo {
    foo?: Record<string, number>
}

const undefinedFoo: OptionalFoo = {
    foo: undefined
    ~~~
};
```

**Comment by issacgerges** [comment]
Ah, thanks for pointing in the direction @fatcerberus. It looks like exactly the gap I believed I found was fixed! I tested against nightly but didn't think to recheck the options. I'll close this one.

**Comment by RyanCavanaugh** [comment]
You can definitely be forgiven for not knowing about a feature that got checked in a week ago! 😅
